lib/rules/no-extend-native.js

```diff
@@ -43,73 +50,127 @@ module.exports = {
     create(context) {
 
         const config = context.options[0] || {};
-        const exceptions = config.exceptions || [];
-        let modifiedBuiltins = Object.keys(globals.builtin).filter(builtin => builtin[0].toUpperCase() === builtin[0]);
+        const exceptions = new Set(config.exceptions || []);
+        const modifiedBuiltins = new Set(
+            Object.keys(globals.builtin)
+                .filter(builtin => builtin[0].toUpperCase() === builtin[0])
+                .filter(builtin => !exceptions.has(builtin))
+        );
+
+        /**
+         * Reports a lint error for the given node.
+         * @param {ASTNode} node The node to report.
+         * @param {string} builtin The name of the native builtin being extended.
+         * @returns {void}
+         */
+        function reportNode(node, builtin) {
+            context.report({
+                node,
+                message: "{{builtin}} prototype is read only, properties should not be added.",
+                data: {
+                    builtin
+                }
+            });
+        }
 
-        if (exceptions.length) {
-            modifiedBuiltins = modifiedBuiltins.filter(builtIn => exceptions.indexOf(builtIn) === -1);
+        /**
+         * Checks that an identifier is an object of a prototype whose member
+         * is being assigned in an AssignmentExpression.
+         * Example: Object.prototype.foo = "bar"
+         * @param {ASTNode} identifierNode The identifier to check.
+         * @returns {boolean} True if the identifier's prototype is modified.
+         */
+        function isInPrototypePropertyAssignment(identifierNode) {
```

Comment:
This function is called after we've already checked that the property in question is `prototype`. That said, the fact that you asked this is a clear indication that the code right now is a little spaghetti and could be made better, so I'll see what I can do on that front.